src/solvers/cma_es.jl

```diff
@@ -0,0 +1,710 @@
+        retract!(M, s.population[i], s.p_m, Y_m, s.σ, s.retraction_method) # Eq. (40)
+    end
+    fitness_vals = map(p -> get_cost(mp, p), s.population)
+    s.best_fitness_current_gen, s.worst_fitness_current_gen = extrema(fitness_vals)
+    s.median_fitness_current_gen = median(fitness_vals)
+    for (i, fitness) in enumerate(fitness_vals)
+        if fitness < s.p_obj
+            s.p_obj = fitness
+            copyto!(s.p, s.population[i])
+        end
+    end
+
+    # sorting solutions
+    ys_c_sorted = map(x -> x[1], sort(collect(zip(s.ys_c, fitness_vals)); by=f -> f[2]))
+
+    # selection and recombination
+    fill!(s.buffer, 0) # from now on until the end of this method buffer is ⟨y⟩_w from Eq. (41)
+    for i in 1:(s.μ) # Eq. (41)
+        s.buffer .+= s.recombination_weights[i] .* ys_c_sorted[i]
+    end
+    new_m = retract(
+        M, s.p_m, get_vector(M, s.p_m, s.buffer, s.basis), s.c_m * s.σ, s.retraction_method
+    ) # Eq. (42)
+
+    # step-size control
+    cinv_y = (cov_invsqrt * s.buffer)
+    s.p_σ .= (1 - s.c_σ) * s.p_σ + sqrt(s.c_σ * (2 - s.c_σ) * s.μ_eff) * cinv_y # Eq. (43)
+    s.σ *= exp(s.c_σ / s.d_σ * ((norm(s.p_σ) / s.e_mv_norm) - 1)) # Eq. (44)
+
+    # covariance matrix adaptation
+    s.p_c .*= 1 - s.c_c # Eq. (45), part 1
+    if norm(s.p_σ) / sqrt(1 - (1 - s.c_σ)^(2 * (iteration + 1))) <
+        (1.4 + 2 / (n_coords + 1)) * s.e_mv_norm # h_σ criterion
+        s.p_c .+= sqrt(s.c_c * (2 - s.c_c) * s.μ_eff) .* s.buffer # Eq. (45), part 2
+        δh_σ = zero(s.c_c) # Appendix A
+    else
+        δh_σ = s.c_c * (2 - s.c_c) # Appendix A
+    end
+    s.covm .*= (1 + s.c_1 * δh_σ - s.c_1 - s.c_μ * sum(s.recombination_weights)) # Eq. (47), part 1  
+    mul!(s.covm, s.p_c, s.p_c', s.c_1, true) # Eq. (47), rank 1 update
+    for i in 1:(s.λ)
+        w_i = s.recombination_weights[i]
+        wᵒi = w_i # Eq. (46)
+        if w_i < 0
+            mul!(cinv_y, cov_invsqrt, s.ys_c[i])
+            wᵒi *= n_coords / norm(cinv_y)^2
+        end
+        mul!(s.covm, s.ys_c[i], s.ys_c[i]', s.c_μ * wᵒi, true) # Eq. (47), rank μ update
+    end
+    # move covariance matrix, p_c and p_σ to new mean point
+    s.covm_eigen = eigen(Symmetric(s.covm))
+    eigenvector_transport!(
+        M, s.covm_eigen, s.p_m, new_m, s.basis, s.vector_transport_method
+    )
+    mul!(
+        s.covm, s.covm_eigen.vectors, Diagonal(s.covm_eigen.values) * s.covm_eigen.vectors'
+    )
+
+    get_vector!(M, Y_m, s.p_m, s.p_σ, s.basis)
+    vector_transport_to!(M, Y_m, s.p_m, Y_m, new_m, s.vector_transport_method)
+    get_coordinates!(M, s.p_σ, new_m, Y_m, s.basis)
+
+    get_vector!(M, Y_m, s.p_m, s.p_c, s.basis)
+    vector_transport_to!(M, Y_m, s.p_m, Y_m, new_m, s.vector_transport_method)
+    get_coordinates!(M, s.p_c, new_m, Y_m, s.basis)
+
+    # update mean point
+    copyto!(M, s.p_m, new_m)
+    return s
+end
+
+function cma_es(M::AbstractManifold, f, p_m=rand(M); kwargs...)
```

Comment:
this one should be documented with all the keyword arguments, because this is the high-level interface the user will usually use.

Also I would prefer if it copies `p` (to some new memory `q` and calls `cam_es!(M,f,q; kwargs...)` with that.

That you already do the interims steps with the cost objective is great!